content_sync/backends/base.py

```diff
@@ -0,0 +1,132 @@
+""" Sync abstract base """
+import abc
+from typing import Any
+
+from content_sync.models import ContentSyncState
+from websites.models import Website
+
+
+class BaseSyncBackend(abc.ABC):
+    """ Base class for syncing backends """
+
+    # NOTE: concrete implementations of this will probably want to
+    #       initialize a client object for the backing service (e.g. ghapi)
+    def __init__(self, website: Website):
+        self.website = website
+
+    @abc.abstractmethod
+    def create_website_in_backend(self):  # pragma: no cover
+        """
+        Called to create the website in the backend.
+
+        An example would be creating a VCS repository.
+        """
+        ...
+
+    @abc.abstractmethod
+    def create_backend_preview(self):  # pragma: no cover
+        """
+        Called to create a site preview in the backend.
+
+        An example would be merging the latest commits into a VCS' preview branch.
+        """
+        ...
+
+    @abc.abstractmethod
+    def create_backend_release(self):  # pragma: no cover
+        """
+        Called to create a site release in the backend.
+
+        An example would be merging the latest commits into a VCS' release branch.
+        """
+        ...
+
+    # NOTE: these next two could perform the same operations if the backing service
+    #       supports an UPSERT type operation (GitHub does)
+    @abc.abstractmethod
+    def create_content_in_backend(
+        self, sync_state: ContentSyncState
+    ):  # pragma: no cover
+        """
+        Called to create a piece of content in the backend.
+
+        An example would be commiting the contents of the WebsiteContent to the VCS repository.
+        """
+        ...
+
+    @abc.abstractmethod
+    def update_content_in_backend(
+        self, sync_state: ContentSyncState
+    ):  # pragma: no cover
+        """
+        Called to create a piece of content in the backend.
+
+        An example would be commiting the updated contents of the WebsiteContent to the VCS repository.
+        """
+        ...
+
+    @abc.abstractmethod
+    def delete_content_in_backend(
+        self, sync_state: ContentSyncState
+    ):  # pragma: no cover
+        """
+        Called to delete a piece of content in the backend.
+
+        An example would be making a commit to delete content in the VCS repository.
+        """
+        ...
+
+    # NOTE: this will be fired for every piece of content that is created, changed, or deleted in the database
+    def sync_content_to_backend(self, sync_state: ContentSyncState):
+        """ Sync a given piece of content given its ContentSyncState """
+        if not sync_state.synced_checksum:
+            self.create_content_in_backend(sync_state)
+        elif sync_state.content is None:
```

Comment:
Does this mean no `ContentSyncState` objects will ever be deleted from the database?  Or will there be a series of steps in deleting content:

1. `ContentSyncState.content` set to `None`
2. `sync_content_to_backend -> delete_content_in_backend` is triggered
3. The `ContentSyncState` object is deleted
4. The `WebsiteContent` object is deleted

Currently, it is not possible to delete a `WebsiteContent` object if a ContentSyncState has a relation to it:
```
"Cannot delete some instances of model 'WebsiteContent' because they are referenced 
through protected foreign keys: 'ContentSyncState.content'."
```

Reply:
Hmmm, yeah, I think I meant to have that relation be a `SET_NULL`, then I changed my mind in the data model PR, forgetting this code, so I'll switch it back over.

Reply:
I set this to `SET_NULL` for now, but need to figure out in more depth how to handle this